sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala

```diff
@@ -89,13 +90,20 @@ case class BroadcastNestedLoopJoinExec(
     }
   }
 
-  @transient private lazy val boundCondition = {
+  private val numMatchedRows = longMetric("numMatchedRows")
```

Comment:
shouldn't this be `lazy`? otherwise `metrics` becomes non lazy...

Reply:
Done. Fixed in latest commit.

Reply:
Actually making it lazy is causing a NullPointerException in SqlQuerySuite, ExistenceJoinSuite, OuterJoinSuite.  Somehow the sqlContext object in line 67 is turning out to be null. I am still trying to understand why that is the case. For now I reverted the change.
`java.lang.NullPointerException
	at org.apache.spark.sql.execution.SparkPlan.sparkContext(SparkPlan.scala:67)
	at org.apache.spark.sql.execution.joins.BroadcastNestedLoopJoinExec.metrics$lzycompute(BroadcastNestedLoopJoinExec.scala:42)
	at org.apache.spark.sql.execution.joins.BroadcastNestedLoopJoinExec.metrics(BroadcastNestedLoopJoinExec.scala:41)
	at org.apache.spark.sql.execution.SparkPlan.longMetric(SparkPlan.scala:135)
	at org.apache.spark.sql.execution.joins.BroadcastNestedLoopJoinExec.numMatchedRows$lzycompute(BroadcastNestedLoopJoinExec.scala:93)